remix-debug/src/debugger/debugger.js

```diff
@@ -0,0 +1,144 @@
+  this.offsetToLineColumnConverter = options.offsetToLineColumnConverter
+  this.compiler = options.compiler
+
+  this.debugger = new Ethdebugger({
+    executionContext: this.executionContext,
+    compilationResult: () => {
+
+      console.dir(self.compiler.lastCompilationResult)
+      window.a = self.compiler.lastCompilationResult
+
+      var compilationResult = this.compiler.lastCompilationResult
+      if (compilationResult) {
+        return compilationResult.data
+      }
+      return null
+    }
+  })
+
+  this.breakPointManager = new remixLib.code.BreakpointManager(this.debugger, (sourceLocation) => {
+    return self.offsetToLineColumnConverter.offsetToLineColumn(sourceLocation, sourceLocation.file, this.compiler.lastCompilationResult.source.sources)
+  }, (step) => {
+    self.event.trigger('breakpointStep', [step])
+  })
+
+  this.debugger.setBreakpointManager(this.breakPointManager)
+
+  this.executionContext.event.register('contextChanged', this, function (context) {
+    // TODO: was already broken
+    // self.switchProvider(context)
+  })
+
+  this.debugger.event.register('newTraceLoaded', this, function () {
+    self.event.trigger('debuggerStatus', [true])
+  })
+
+  this.debugger.event.register('traceUnloaded', this, function () {
+    self.event.trigger('debuggerStatus', [false])
+  })
+
+  this.event.register('breakpointStep', function (step) {
+    self.step_manager.jumpTo(step)
+  })
+
+  this.debugger.addProvider('vm', this.executionContext.vm())
+  this.debugger.addProvider('injected', this.executionContext.internalWeb3())
+  this.debugger.addProvider('web3', this.executionContext.internalWeb3())
+  this.debugger.switchProvider(this.executionContext.getProvider())
+}
+
+Debugger.prototype.registerAndHighlightCodeItem = function (index) {
+  const self = this
+  // register selected code item, highlight the corresponding source location
+  if (!self.compiler.lastCompilationResult) return
+  self.debugger.traceManager.getCurrentCalledAddressAt(index, (error, address) => {
+    if (error) return console.log(error)
+    self.debugger.callTree.sourceLocationTracker.getSourceLocationFromVMTraceIndex(address, index, self.compiler.lastCompilationResult.data.contracts, function (error, rawLocation) {
+      if (!error && self.compiler.lastCompilationResult && self.compiler.lastCompilationResult.data) {
+        var lineColumnPos = self.offsetToLineColumnConverter.offsetToLineColumn(rawLocation, rawLocation.file, self.compiler.lastCompilationResult.source.sources)
+        self.event.trigger('newSourceLocation', [lineColumnPos, rawLocation])
+      } else {
+        self.event.trigger('newSourceLocation', [null])
+      }
+    })
+  })
+}
+
+Debugger.prototype.debug = function (blockNumber, txNumber, tx, loadingCb) {
+  const self = this
+  let web3 = this.executionContext.web3()
+
+  if (this.debugger.traceManager.isLoading) {
+    return
+  }
+
+  self.debugger.solidityProxy.reset({})
```

Comment:
I think clearing this now will make the soucehighlighting and breakpoint not available.

Reply:
this seems to be done in https://github.com/ethereum/remix/blob/master/remix-debug/src/Ethdebugger.js#L223